### ISSUE TITLE
ogbg fails on pytorch submissions, pthread_create() failed

### ISSUE THREAD

**Issue by tfaod** [mlcommons/algorithmic-efficiency#866]
This high-priority bug is blocking all submissions on the self-tuning leaderboard. We cannot score any Pytorch optimizers before this is fixed.

## Description
ogbg workloads are consistently failing on Pytorch submissions. This includes:
* self-tuning submissions: all runs crash for nadamw, schedulefree_v1/v2 [link](https://github.com/mlcommons/submissions_algorithms/blob/76f150fdecf1facf226b88e1765f0121016e5296/logs/self_tuning/schedule_free_adamw_v2/study_0/ogbg_pytorch/ogbg_pytorch_03-18-2025-15-57-12.log#L807
)
* external-tuning submissions: most shampoo runs crash [link](https://github.com/mlcommons/submissions_algorithms/blob/main/logs/external_tuning/shampoo/study_0/ogbg_pytorch/ogbg_pytorch_03-15-2025-19-16-26.log), except for one set of hyperparameters [link](https://github.com/mlcommons/submissions_algorithms/blob/main/logs/external_tuning/shampoo/study_0/ogbg_pytorch/ogbg_pytorch_03-15-2025-19-17-10.log)

Main error message:
```
2025-03-18 17:26:43.777726: F external/tsl/tsl/platform/default/env.cc:74] Check failed: ret == 0 (11 vs. 0)Thread tf_XLACpuCompile creation via pthread_create() failed.
Fatal Python error: Aborted
```

When we run ogbg workloads with Pytorch, we notice that metric computations leverage a clu metric. This  calls a jax package dependency which subsequently leads to an XLA error. We suspect that this crash happens when we reach some thread or memory limit in the jax computation. The runs (in both self and external tuning) crash consistently around 15,000 training steps. 

When checking external tuning Pytorch runs, we observe that this error appeared on **all but one hyperparameter set** in the Pytorch Distributed Shampoo runs. Interestingly, the runs which did not crash reached the target well before the 15,000 training step threshold, whereas the remaining other runs crashed after reaching the threshold.

```
...
  File "/usr/local/lib/python3.11/site-packages/jax/_src/interpreters/pxla.py", line 2363 in compile
...
  File "/usr/local/lib/python3.11/site-packages/jax/_src/core.py", line 387 in bind
  File "/usr/local/lib/python3.11/site-packages/jax/_src/lax/lax.py", line 420 in add
  File "/usr/local/lib/python3.11/site-packages/jax/_src/numpy/lax_numpy.py", line 5377 in _normalize_index
  File "/usr/local/lib/python3.11/site-packages/jax/_src/numpy/lax_numpy.py", line 5767 in <genexpr>
  File "/usr/local/lib/python3.11/site-packages/jax/_src/numpy/lax_numpy.py", line 5769 in _index_to_gather
  File "/usr/local/lib/python3.11/site-packages/jax/_src/numpy/lax_numpy.py", line 5625 in _gather
  File "/usr/local/lib/python3.11/site-packages/jax/_src/numpy/lax_numpy.py", line 5616 in _rewriting_take
  File "/usr/local/lib/python3.11/site-packages/jax/_src/array.py", line 336 in __getitem__
  File "/algorithmic-efficiency/algoperf/workloads/ogbg/metrics.py", line 65 in compute
  File "/usr/local/lib/python3.11/site-packages/clu/metrics.py", line 626 in <dictcomp>
  File "/usr/local/lib/python3.11/site-packages/clu/metrics.py", line 625 in compute
  File "/algorithmic-efficiency/algoperf/workloads/ogbg/ogbg_pytorch/workload.py", line 243 in _normalize_eval_metrics
  File "/algorithmic-efficiency/algoperf/workloads/ogbg/workload.py", line 193 in _eval_model_on_split
...
```

## Steps to Reproduce
* Run ogbg with Pytorch :) 

## Debugging Ideas

1. Try increasing the number of threads (system limit)
  * David has tried increasing system limit?? (unclear what it was increased to, try other values?)
2. Try downgrading to algoperf v1 packages
* Downgrade clu/jax to previous package versions, where ogbg was running correctly
3. Ask for more help
* clu repository https://github.com/google/CommonLoopUtils/blob/main/clu/metrics.py 
* Reaching out to clu or Pytorch teams. If a new version of pytorch/clu is crashing our runs by requiring more threads or memory than the previous versions, this is an inefficiency in their code that they should fix.

**Comment by davidtweedle** [comment]
I was able to avoid this thread error on one test run so far (To confirm I will run schedule free next).
Using a dummy optimizer, I was able to run ogbg on pytorch for >3 hours (50k training steps). The dummy optimizer fails in the same manner as others without the fix.
The problem seems to be with jax compilation on cpu combined with torchrun --nproc_per_node=8.
To fix, the jax arrays should be converted to numpy arrays as soon as possible when using pytorch ddp.

**Comment by davidtweedle** [comment]
You can see here that schedule free now completes the run with the changes.
https://pastebin.com/RgqMqgkb
